### DISCUSSION TITLE
Apply Konsist tests to a specific set of files (i.e. git diff)

### DISCUSSION THREAD

**Discussion by erawhctim** [LemonAppDev/konsist#649] [Request a Feature]
### Tell us more.

There are a few testing scenarios we can imagine for our project that we'd love to enforce via Konsist, but some tests we only want to enforce on newly-added code (vs. retroactively applying them to the entire project). As an example, I'd love to enforce that a certain file type _always_ includes a test when it's added to the project (but I can't write a Konsist test to enforce that without also ensuring that existing/legacy code follows that same rule). 

It'd be great if Konsist had a separate scope concept based on a git diff (or an arbitrary set of file paths) so that we could run a subset of Konsist tests and only apply them to those specific files. Is something like that even possible?

**Comment by igorwojda** [comment]
The new upcoming version will have better support for creating declarations from a single file. 
e.g. here JGit was used:
See https://github.com/LemonAppDev/konsist/discussions/586

We are exploring this scenario

**Reply by erawhctim** [reply]
nice! this should definitely help with my use case, thanks

**Comment by igorwojda** [comment]
A new version v0.13.0 has been released. It contains the fix.
It allows to creation of KoScope from the list of files.
See https://github.com/LemonAppDev/konsist/releases/tag/v0.13.0

```kotlin
// Create scope from paths
Konsist.scopeFromFiles("Repository.kt", "path\UseCase.kt")

// Create scope from set of paths
val paths = setOf("Repository.kt", "path\UseCase.kt")
Konsist.scopeFromFiles(paths)
```

At some point, I will add it to the docs, but I need to experiment with this first. Here is the high-level overview:

Refactoring per module is a recommended way to go, however, some opt for a distinct strategy. Rather than refactoring and incorporating Konsist tests in module one, they emphasize ensuring that the newly incorporated Kotlin files are properly structured. Konsist facilitates this flow, although this requires a “multi-level” configuration. Here is the high-level overview:

1. Get the list of changed files in CI
2. Pass the list of changed files to the build system
3. Pass the list of changed files to the test source set
4. Retrieve a list of changed files in the test
5. Create Konsist scope from the list of changed files
6. Define Konsist checks


1. The [chaged-files](https://github.com/tj-actions/changed-files) GitHub Action allows to retrieve the list of changed files:
```
- name: Get Changed Files
        id: changed-files
        uses: tj-actions/changed-files@v39.2.0

- name: List Changed Files
        run: echo Changed files: '\n' ${{ steps.changed-files.outputs.all_changed_files }} | tr ' ' '\n'
```

...
